Add http monitoring for windows via ETW HttpService provider

derekwbrown opened the pull request:
Initial revision.
All windows HTTP changes on top of latest main.

### What does this PR do?

Adds support for HTTP stats for windows in NPM.
Adds support for HTTPS stats when using IIS.



### Possible Drawbacks / Trade-offs


### Describe how to test/QA your changes


### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

iglendd committed:
Add http monitoring for windows via ETW HttpService provider

Initial revision.
All windows HTTP changes on top of latest main.

WIP -- make linux build

linux compiles; back to windows

windows compiles too

appears to be working

More WIP:

fix path parsing

:wqmore spelling changes

get past source test

iglendd committed:
Add http monitoring for windows via ETW HttpService provider

Initial revision.
All windows HTTP changes on top of latest main.

juliogreff committed:
[ad] Wait for first workloadmeta bundle on provider start

This fixes an issue in `agent check`, where a subtle change in timing
introduced in #12968 meant that configs did not get scheduled in time
for `agent check` to see them, effectively appearing as if they were not
present. This change ensures that `kubelet` and `container` providers
only return from `Collect` if they've received at least one event bundle
from workloadmeta.

This also changes workloadmeta to actually allow the sending of empty
bundles, as that's necessary to signal that the subscription has started
and there are indeed no entities in the store. Otherwise, config
providers in non-containerized environments will hang indefinitely.

juliogreff committed:
[ad] Generate configs only once for k8s containers

Previously, we'd generate configs for the container events themselves,
and again when handling the pod event. When the container belonged to an
AD-annotated pod, it would briefly have a `container_collect_all` config
before honoring the one in the annotations. Now, since the configs are
generated entirely at once, AutoDiscovery can ignore the
`container_collect_all` correctly the first time.

iglendd reviewed (approved):
Looks good (mostly looking from the perspective of fidelity of combining two pending branches)

derekwbrown committed:
WIP -- make linux build

derekwbrown committed:
linux compiles; back to windows

derekwbrown committed:
First pass convert linux http transaction to interface

derekwbrown committed:
windows compiles too

derekwbrown committed:
appears to be working

More WIP:

fix path parsing

Add http monitoring for windows via ETW HttpService provider

Initial revision.
All windows HTTP changes on top of latest main.

WIP -- make linux build

linux compiles; back to windows

First pass convert linux http transaction to interface

appears to be working

More WIP:

fix path parsing

fix spelling mistake

juliogreff committed:
[ad] Add container_collect_all after annotation configs

ISauve commented on pkg/network/encoding/http.go:
```diff
@@ -97,13 +99,14 @@ func newHTTPEncoder(payload *network.Connections) *httpEncoder {
+		return nil, 0, nil
 	}
 
 	keyTuple := network.HTTPKeyTupleFromConn(c)
```
The heuristic used by this function to determine which side of the connection is the client & which is the server  doesn't work on windows, so we need to modify this function to return both possible keys & check for aggregations for both keys: see https://github.com/DataDog/datadog-agent/blob/aeb4e8d7a1a1c73c165bda6401a051c1be4fe3dd/pkg/network/event_common.go#L383-L396 and https://github.com/DataDog/datadog-agent/blob/aeb4e8d7a1a1c73c165bda6401a051c1be4fe3dd/pkg/network/encoding/http.go#L61-L66

ISauve commented on pkg/network/http/batch_manager.go:
```diff
@@ -84,7 +84,11 @@ func (m *batchManager) GetTransactionsFrom(notification httpNotification) ([]htt
 	state.pos = 0
 
-	return batch.Transactions()[offset:], nil
+	txns := make([]httpTX, len(batch.Transactions()[offset:]))
+	for idx, tx := range batch.Transactions()[offset:] {
+		txns[idx] = &tx
+	}
+	return txns, nil
```
This change seems like it may have been accidentally included? (Since it's modifying ebpf code)

ISauve commented on pkg/network/http/driver_interface.go:
```diff
@@ -0,0 +1,178 @@
+			break
+		}
+		count += len(txns)
+		di.dataChannel <- txns
```
Should we remove this channel and have this function read all the transactions available, return them as a batch & then we can process them synchronously? This should help mitigate the race condition which you mentioned

ISauve commented on pkg/network/http/etw_interface.go:
```diff
@@ -0,0 +1,124 @@
+func newHttpEtwInterface(c *config.Config) *httpEtwInterface {
+	return &httpEtwInterface{
+		maxEntriesBuffered: c.MaxHTTPStatsBuffered,
+		dataChannel:        make(chan []etw.Http),
```
I think this suffers from the same race condition as the driver interface - we should consider removing this channel & processing synchronously here as well

derekwbrown replied:
@ISauve For now, I'm going to leave this like this.
I have a follow-on PR that does some better converging of the linux/windows.
And then I want to re-examine how we do this, and whether the channel is necessary.
If nothing else, it will need some different locking, but I don't want to make this PR more crazy than it already is.

ISauve replied:
Sounds good to me!

ISauve commented on pkg/network/http/monitor_etw_windows.go:
```diff
@@ -0,0 +1,133 @@
+	m.statkeeper.Process(transactions)
+}
+
+func (m *EtwMonitor) removeDuplicates(stats map[Key]*RequestStats) {
+	// With localhost traffic, the driver will create a flow for both endpoints. Both
+	// these flows will be normalized so that source=client and dest=server, which
+	// results in 2 identical http transactions being sent up to userspace & processed.
+	// To fix this, we'll find all localhost keys and half their transaction counts.
+
+	for k, v := range stats {
+		if isLocalhost(k) {
+			v.HalfAllCounts()
+		}
+	}
+}
```
Is this true for ETW as well as the driver?

ISauve replied:
Did we double check this? If it's not true for ETW, we'll end up with data correctness issues

ISauve replied:
@iglendd could you verify?

iglendd replied:
To be honest I have copied the code without understanding of its implication. The comment is certainly wrong and the function is I think is not necessary for ETW.

ISauve commented on pkg/network/http/model_linux.go:
```diff
@@ -0,0 +1,198 @@
+// GET variables excluded.
+// Example:
+// For a request fragment "GET /foo?var=bar HTTP/1.1", this method will return "/foo"
+func (tx *ebpfHttpTx) Path(buffer []byte) ([]byte, bool) {
```
Rather than have this function be defined in multiple places, I think we should remove this function from the http transaction interface, replace it with `ReqFragment() []byte` and have the http statkeeper handle extracting the path from the fragment (using the `getPath` function)

ISauve replied:
@derekwbrown is this one of the changes which your follow-up PR is going to address?

derekwbrown replied:
it should ;)

ISauve commented on pkg/network/http/model_linux.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+func (tx *ebpfHttpTx) ReqFragment() []byte {
+	asslice := (*[1 << 30]byte)(unsafe.Pointer(&tx.request_fragment))[:int(C.HTTP_BUFFER_SIZE):int(C.HTTP_BUFFER_SIZE)]
```
camelcase might be good for this variable name... 🤣

ISauve commented on pkg/network/config/config.go:
```diff
@@ -201,9 +204,10 @@ func New() *Config {
-		MaxHTTPStatsBuffered:  100000,
+		EnableHTTPMonitoring:            cfg.GetBool(join(netNS, "enable_http_monitoring")),
+		EnableHTTPSMonitoring:           cfg.GetBool(join(netNS, "enable_https_monitoring")),
+		EnableHTTPHTTPSMonitoringViaETW: cfg.GetBool(join(netNS, "enable_http_https_monitoring_via_etw")),
```
I think if `EnableHTTPHTTPSMonitoringViaETW` is enabled, that should imply that http monitoring is enabled, so we should set those configuration values accordingly

ISauve commented on pkg/network/tracer/tracer_windows.go:
```diff
@@ -218,3 +221,25 @@ func (t *Tracer) DebugCachedConntrack(ctx context.Context) (interface{}, error)
 }
+
+func newHttpMonitor(c *config.Config, dh driver.Handle) http.Monitor {
+	if !c.EnableHTTPMonitoring {
+		return nil
+	}
```
As mentioned above, currently having `EnableHTTPHTTPSMonitoringViaETW` enabled doesn't do anything unless http monitoring is also explicitly enabled - I think enabling etw monitoring shouldn't require 2 configuration values

ahmed-mez committed:
[auto-instru] Fix NODE_OPTIONS value (#13122)

juliogreff committed:
[ad] Clarify code comments

juliogreff committed:
[ad] Send empty changesets from KubeContainerConfigProvider

The config poller expects a changeset to be able to mark a streaming
provider as started. Otherwise, `AutoConfig.LoadAndRun` will block,
potentially forever, breaking AD.

juliogreff committed:
[ad] Preserve order in generated configs

This is a follow up to an earlier commit in this PR, that attempted to
generate container_collect_all configs after all others. That commit
missed the mangling done by `processEvents`. Now it does preserve order
correctly.

ahmed-mez committed:
[inv/tasks] Fix cluster agent ldflags (#13124)

scottopell committed:
Print status output directly to stdout without 'logger' prefixes (#13117)

* Use scrubber directly to print things to stdout without logger formatting

* Adds release notes for this change in behavior

* Update releasenotes/notes/plain-json-agent-status-5d0a427ae6cc2338.yaml

Co-authored-by: May Lee <mayl@alumni.cmu.edu>

* Update releasenotes/notes/plain-json-agent-status-5d0a427ae6cc2338.yaml

Co-authored-by: May Lee <mayl@alumni.cmu.edu>

* Update releasenotes/notes/plain-json-agent-status-5d0a427ae6cc2338.yaml

Co-authored-by: May Lee <mayl@alumni.cmu.edu>

Co-authored-by: May Lee <mayl@alumni.cmu.edu>

derekwbrown committed:
more test fixes

derekwbrown committed:
fix more tests

derekwbrown committed:
more test fixin

derekwbrown committed:
Fix problem not copying http transactions from the buffer
to the local statkeeper

derekwbrown committed:
remove extra unnecessary copy

derekwbrown committed:
Fix crash when http _not_ enabled

derekwbrown committed:
fix crash

derekwbrown committed:
Merge branch 'main' into db/merged-http-windows

derekwbrown committed:
fix merge

derekwbrown committed:
ci fixes

derekwbrown committed:
Remove unused import

djmitche commented on cmd/agent/app/status.go:
```diff
@@ -156,6 +156,8 @@ func requestStatus() error {
 	}
 
+	s = scrubMessage(s)
+
```
This is double-scrubbing non-JSON output (see three lines earlier).

djmitche commented on cmd/agent/app/status.go:
```diff
@@ -211,6 +213,8 @@ func componentStatus(component string) error {
 		s = scrubMessage(string(r))
 	}
 
+	s = scrubMessage(s)
```
Same here

djmitche commented on pkg/metadata/inventories/README.md:
```diff
@@ -93,6 +93,7 @@ The payload is a JSON dict with the following fields
     `network_config.enabled` config option in `system-probe.yaml`).
   - `feature_networks_http_enabled` - **bool**: True if HTTP monitoring is enabled for Network Performance Monitoring (see: `network_config.enable_http_monitoring` config option in `system-proble.yaml`).
   - `feature_networks_https_enabled` - **bool**: True if HTTPS monitoring is enabled for Network Performance Monitoring (see: `network_config.enable_https_monitoring` config option in `system-proble.yaml`).
+  - `feature_networks_https_enabled` - **bool**: True if HTTP or HTTPS monitoring on Windows via ETW is enabled for Network Performance Monitoring (see: `network_config.enable_http_https_via_etw_monitoring` config option in `system-proble.yaml`).
```
```suggestion
  - `feature_networks_http_https_via_etw_enabled` - **bool**: True if HTTP or HTTPS monitoring on Windows via ETW is enabled for Network Performance Monitoring (see: `network_config.enable_http_https_via_etw_monitoring` config option in `system-proble.yaml`).
```
it'd also be good to spell out, or linkify, ETW. I have no idea what that means :)

derekwbrown committed:
review feedback

derekwbrown committed:
resolve https/etw doc problem discovered in review

derekwbrown committed:
Merge branch 'main' into db/merged-http-windows

derekwbrown committed:
redo conntuple code, fix merge error

derekwbrown commented on pkg/network/dns/driver_windows.go:
```diff
@@ -16,10 +16,9 @@ import (
 	"time"
 	"unsafe"
 
+	"github.com/DataDog/datadog-agent/pkg/network/driver"
```
I think it's technically right.  Alphabetically.

iglendd replied:
I have seen other companies have tree section in imports: global, github, private. I thought that Datadog Agent code did it as well, hence my comment.

djmitche replied:
We do, but we don't enforce it with an automatic linter, since that didn't work very well.  See https://github.com/DataDog/datadog-agent/blob/main/docs/dev/imports.md

derekwbrown committed:
review feedback

derekwbrown committed:
fix linux build error

derekwbrown committed:
gofmt

djmitche reviewed (approved):
Thanks for the fixes!

derekwbrown committed:
Remove `removeDuplicates()` on ETW; was agreed that it's unnecessary

derekwbrown committed:
Merge branch 'main' into db/merged-http-windows

derekwbrown committed:
gofmt

derekwbrown committed:
Merge branch 'main' into db/merged-http-windows

derekwbrown committed:
fix merge problems (again)

brycekahle committed:
Fix usage of inet.af/netaddr to use net/netip
